### PR TITLE
Overview page add state for missing data source

### PR DESCRIPTION
**Pull request by TackAdam**
### Description
Add a state for workspaces when there is no data source associated. A data source needs to be present for everything to function so the user is prompted to do so if they end up in a state where a data source is not present. 

https://github.com/user-attachments/assets/24d5129d-8dec-47f0-9b48-6faf1b3a2a6f

### Check List
- [x] New functionality includes testing.
  - [] All tests pass, including unit test, integration test and doctest
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
  - [ ] New functionality has user manual doc added
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: common/constants/shared.ts [hunk 1]
```diff
@@ -35,6 +35,7 @@ export const JOBS_ENDPOINT_BASE = '/_plugins/_async_query';
 export const JOB_RESULT_ENDPOINT = '/result';
 
 export const tutorialSampleDataPluginId = 'import_sample_data';
+export const dataSourcesId = 'dataSources';
```
Comment (ps48) [inline common/constants/shared.ts:38]: Can we please name this as `dataSourceManagementPluginId`?

File: public/components/overview/components/dashboard_controls.tsx [hunk 2]
```diff
@@ -13,19 +14,56 @@ import {
...
+      const workspaceId = getWorkspaceIdFromUrl(currentUrl, coreRefs?.http?.basePath.getBasePath());
+
+      if (workspaceId) {
+        setIsInWorkspace(true);
+        const savedObjectsArray = await getDatasourceAttributes();
+        setIsDataSourceEmpty(savedObjectsArray.length === 0);
+
+        // Set to null if there are no data sources associated [Handle if dashboard was set, then datasource deleted]
+        if (savedObjectsArray.length === 0) {
+          await setObservabilityDashboardsId(null);
```
Comment (ps48) [inline public/components/overview/components/dashboard_controls.tsx:57]: For the refresh problem, I looked into the CMS implementation. It looks like this is an issue from the CMS plugin where the registration is only happening in plugin setup: `https://github.com/opensearch-project/dashboards-observability/blob/6a43f539745ad1d51d66d99962563c620b681e4b/public/plugin.tsx#L192` May be we can add a call-back for this function. This callback can be called to run `setupOverviewPage`

File: public/components/overview/components/dashboard_controls.tsx [hunk 3]
```diff
@@ -13,19 +14,56 @@ import {
...
+};
 
 export function DashboardControls() {
+  const [isDataSourceEmpty, setIsDataSourceEmpty] = useState<boolean | null>(null);
+  const [isInWorkspace, setIsInWorkspace] = useState<boolean>(false); // Track if user is in a workspace
   const isDashboardSelected = useObservable(ObsDashboardStateManager.isDashboardSelected$);
   const dashboardState = useObservable(ObsDashboardStateManager.dashboardState$);
 
+  useEffect(() => {
+    const checkDataSource = async () => {
```
Comment (ps48) [inline public/components/overview/components/dashboard_controls.tsx:46]: Let's add a unit test for this function.

File: common/constants/shared.ts (hunk 1 above)
Comment (TackAdam) [reply common/constants/shared.ts:38]: Changed.

File: public/components/overview/components/dashboard_controls.tsx (hunk 2 above)
Comment (TackAdam) [reply public/components/overview/components/dashboard_controls.tsx:57]: Screenshot 2024-11-11 at 10 46 31 AM
Wasn't able to use a callback as the registration exist. Found a work around calling
 
getOverviewPage().removeSection(SECTIONS.DASHBOARD);

to remove the section getting it to re-render correctly.

File: public/components/overview/components/dashboard_controls.tsx (hunk 3 above)
Comment (TackAdam) [reply public/components/overview/components/dashboard_controls.tsx:46]: Added

File: public/components/overview/components/add_datasource_callout.tsx [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        </EuiFlexItem>
+        <EuiFlexItem grow={false}>
+          <EuiTitle size="m">
+            <h3>No connected data sources</h3>
+          </EuiTitle>
+        </EuiFlexItem>
+        <EuiFlexItem grow={false} style={{ maxWidth: '40%' }}>
+          <EuiText textAlign="center">
+            <p style={{ margin: 0 }}>
+              There are no data sources associated to the workspace. Associate data sources or
```
Comment (mengweieric) [inline public/components/overview/components/add_datasource_callout.tsx:34]: minor: add i18n please (can be followups)

File: public/components/overview/home.tsx [hunk 5]
```diff
@@ -92,6 +97,7 @@ export const Home = () => {
             startDate: dashboardAttributes.timeFrom,
             endDate: dashboardAttributes.timeTo,
           };
+          setIsDashboardLoaded(true);
```
Comment (mengweieric) [inline public/components/overview/home.tsx:100]: we should be able to set this flag in finally block

File: public/components/overview/components/dashboard_controls.tsx [hunk 6]
```diff
@@ -13,19 +14,58 @@ import {
...
```
Comment (mengweieric) [inline public/components/overview/components/dashboard_controls.tsx:47]: We should move this function definition out of the useEffect to make it clearer

File: public/components/overview/components/select_dashboard_flyout.tsx [hunk 7]
```diff
@@ -61,6 +63,7 @@ export function SelectDashboardFlyout({ closeFlyout, dashboardsSavedObjects, rel
   const onClickAdd = async () => {
     const selectedOption = options.find((option) => option.checked === 'on');
     if (selectedOption && selectedOption.key) {
+      getOverviewPage().createSection(DASHBOARD_SECTION);
```
Comment (ps48) [inline public/components/overview/components/select_dashboard_flyout.tsx:66]: Does replace dashboard work as expected when the section is already created?

File: public/components/overview/components/add_dashboard_callout.tsx [hunk 8]
```diff
@@ -26,7 +26,7 @@ import SampleDataLightPNG from './assets/SampleDataLight.png';
 
 export function AddDashboardCallout() {
   const showFlyout = useObservable(ObsDashboardStateManager.showFlyout$);
-  const isDarkMode = uiSettingsService.get('theme:darkMode');
+  const isDarkMode = uiSettingsService?.get('theme:darkMode') ?? false;
```
Comment (ps48) [inline public/components/overview/components/add_dashboard_callout.tsx:29]: nice

**Review by ps48 — APPROVED** [review]

File: public/components/overview/components/select_dashboard_flyout.tsx (hunk 7 above)
Comment (TackAdam) [reply public/components/overview/components/select_dashboard_flyout.tsx:66]: Yes, haven't seen any problems in all my testing. Just added as a fail safe for if a user removed the datasource to ensure when it is added back it works without needing a hard refresh.

File: public/components/overview/components/add_datasource_callout.tsx (hunk 4 above)
Comment (TackAdam) [reply public/components/overview/components/add_datasource_callout.tsx]: added

              {i18n.translate('traceAnalytics.noDataSourcesMessage', {
                defaultMessage:
                  'There are no data sources associated to the workspace. Associate data sources or request your administrator to associate data sources for you to get started.',
              })}

File: public/components/overview/components/dashboard_controls.tsx (hunk 6 above)
Comment (TackAdam) [reply public/components/overview/components/dashboard_controls.tsx]: Done

**Review by mengweieric — APPROVED** [review]
Thanks

**Pull request merged**